### ISSUE TITLE
Implementing ISE through Hugging Face Transformers

### ISSUE THREAD

**Issue by lindsey98** [tongwu2020/ISE#2]
Hi Tong, I noticed that your LLaMA model is implemented using the Torchtune framework. I'm currently trying to replicate it using the Hugging Face Transformers library, as I’d like to integrate it into the StruQ and SecAlign training code.

May I confirm if the following implementation aligns with your idea? In my version, ``expert_labels`` represents segment IDs (i.e., 0, 1, 2, 3), and I apply a shift based on these labels to be added to the word embeddings.

Thanks for your time!
```
class LlamaModel(transformers.LlamaModel):
    """
    Transformer decoder consisting of *config.num_hidden_layers* layers. Each layer is a [`LlamaDecoderLayer`]

    Args:
        config: LlamaConfig
    """

    def __init__(self, config: LlamaMoEConfig):
        super().__init__(config)
        del self.layers
        self.layers = nn.ModuleList(
            [LlamaDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
        )
        self.apply_input_shifts = config.apply_input_shifts
        self.input_shifts = nn.Embedding(config.num_experts, config.hidden_size) # config.num_experts=4

    def forward(
        self,
        input_ids: torch.LongTensor = None,
        expert_labels: torch.LongTensor = None,
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_values: Optional[Union[Cache, List[torch.FloatTensor]]] = None,
        inputs_embeds: Optional[torch.FloatTensor] = None,
        use_cache: Optional[bool] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
        cache_position: Optional[torch.LongTensor] = None,
    ) -> Union[Tuple, BaseModelOutputWithPast]:

        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        use_cache = use_cache if use_cache is not None else self.config.use_cache
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        if (input_ids is None) ^ (inputs_embeds is not None):
            raise ValueError(
                "You cannot specify both input_ids and inputs_embeds at the same time, and must specify either one"
            )

        if self.gradient_checkpointing and self.training and use_cache:
            logger.warning_once(
                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
            )
            use_cache = False

        if inputs_embeds is None:
            inputs_embeds = self.embed_tokens(input_ids)

        ......

        causal_mask = self._update_causal_mask(
            attention_mask, inputs_embeds, cache_position, past_key_values, output_attentions
        )
        hidden_states = inputs_embeds

        ## apply ISE
        if self.apply_input_shifts:
            shifts = self.input_shifts(expert_labels)  # (batch_size, seq_len, hidden_size)
            hidden_states = hidden_states + shifts

        # create position embeddings to be shared across the decoder layers
        position_embeddings = self.rotary_emb(hidden_states, position_ids)

        # decoder layers
        all_hidden_states = () if output_hidden_states else None
        all_self_attns = () if output_attentions else None
        next_decoder_cache = None

        for decoder_layer in self.layers:
            if output_hidden_states:
                all_hidden_states += (hidden_states,)

            if self.gradient_checkpointing and self.training:
                layer_outputs = self._gradient_checkpointing_func(
                    decoder_layer.__call__,
                    hidden_states,
                    expert_labels,
                    causal_mask,
                    position_ids,
                    past_key_values,
                    output_attentions,
                    use_cache,
                    cache_position,
                    position_embeddings,
                )
            else:
                layer_outputs = decoder_layer(
                    hidden_states,
                    expert_labels=expert_labels,
                    attention_mask=causal_mask,
                    position_ids=position_ids,
                    past_key_value=past_key_values,
                    output_attentions=output_attentions,
                    use_cache=use_cache,
                    cache_position=cache_position,
                    position_embeddings=position_embeddings,
                )

            hidden_states = layer_outputs[0]

           ......
```

**Comment by tongwu2020** [comment]
Hi, I think the current code looks good. One improvement you could make is adding a tunable initialization function for `self.input_shifts = nn.Embedding(config.num_experts, config.hidden_size)`, allowing you to control how the embedding is initialized with different mean/std.

**Comment by lindsey98** [comment]
sure, thank you!
